libraries/chain/resource_limits.cpp

```diff
@@ -220,7 +223,9 @@ void resource_limits_manager::add_pending_ram_usage( const account_name account,
    _db.modify( usage, [&]( auto& u ) {
       u.ram_usage += ram_delta;
 
-      if (auto dm_logger = _get_deep_mind_logger()) {
+      // add_pending_ram_usage is only called by transaction_context::schedule_transaction,
```

Comment:
This comment isn't accurate. There are other ways `add_pending_ram_usage` can be called. It needs to have a `bool is_trx_transient` parameter plumbed in as well. In most places it is called, `false` can just be passed in as an argument for that new parameter. The only exception is from within `transaction_context::add_ram_usage` where `is_transient()` should be used instead.

Note that `transaction_context::add_ram_usage` is only directly called from `transaction_context::schedule_transaction` and `apply_context::add_ram_usage`. But that second usage of it is what allows it to be indirectly called from a bunch of places within `apply_context.cpp` (DB operations, and scheduling/cancelling deferred transactions) and `eosio_context.cpp` (native actions). And while DB operations and native operations are disallowed in read-only transactions, they can still be used by dry-run transactions which are considered transient.

Also note that scheduling/canceling deferred transactions also needs to be restricted from within a read-only transaction context.